agent/consul/gateways/controller_gateways.go

```diff
@@ -840,151 +828,132 @@ func newGatewayMeta(gateway *structs.APIGatewayConfigEntry, bound structs.Config
 	}).initialize()
 }
 
-// gatewayConditionGenerator is a simple struct used for isolating
-// the status conditions that we generate for our components
-type gatewayConditionGenerator struct {
-	now *time.Time
-}
-
-// newGatewayConditionGenerator initializes a status conditions generator
-func newGatewayConditionGenerator() *gatewayConditionGenerator {
-	return &gatewayConditionGenerator{
-		now: pointerTo(time.Now().UTC()),
-	}
+// gatewayAccepted marks the APIGateway as valid.
+func gatewayAccepted() structs.Condition {
+	return structs.NewGatewayCondition(
+		structs.GatewayConditionAccepted,
+		structs.ConditionStatusTrue,
+		structs.GatewayReasonAccepted,
+		"gateway is valid",
+		structs.ResourceReference{},
+	)
 }
 
 // invalidCertificate returns a condition used when a gateway references a
 // certificate that does not exist. It takes a ref used to scope the condition
 // to a given APIGateway listener.
-func (g *gatewayConditionGenerator) invalidCertificate(ref structs.ResourceReference, err error) structs.Condition {
-	return structs.Condition{
-		Type:               "Accepted",
-		Status:             "False",
-		Reason:             "InvalidCertificate",
-		Message:            err.Error(),
-		Resource:           pointerTo(ref),
-		LastTransitionTime: g.now,
-	}
+func invalidCertificate(ref structs.ResourceReference, err error) structs.Condition {
+	return structs.NewGatewayCondition(
+		structs.GatewayConditionResolvedRefs,
+		structs.ConditionStatusFalse,
+		structs.GatewayListenerReasonInvalidCertificateRef,
+		err.Error(),
+		ref,
+	)
 }
 
 // invalidCertificates is used to set the overall condition of the APIGateway
 // to invalid due to missing certificates that it references.
-func (g *gatewayConditionGenerator) invalidCertificates() structs.Condition {
-	return structs.Condition{
-		Type:               "Accepted",
-		Status:             "False",
-		Reason:             "InvalidCertificates",
-		Message:            "gateway references invalid certificates",
-		LastTransitionTime: g.now,
-	}
+func invalidCertificates() structs.Condition {
```

Comment:
now this and `invalidCertificate` fail for the same Condition/Reason but have different messages

Reply:
so @andrewstucki this is now the only one (along with `invalidCertificate` ) to have a different type/reason (so `ResolvedRefs` type and `InvalidCertifcateRef` for the reason vs `Accepted` type and `InvalidCertificate` for the reason). Do you have an opinion on whether this makes sense or should be reverted?

Reply:
So, I'm thinking that it makes sense to move `invalidCertificate` to a `ResolvedRefs` status along with keeping it scoped to an individual listener. I'd be _ok_ potentially dropping this entire status, but maybe in a separate PR after discussing a bit more. Specifically here's the issue:

1. I believe we want `Accepted` to be a rolled up condition as to whether a Gateway/Route is invalid or not
2. We actually have a check that a Gateway has `Accepted` == `True` in order for it to actually bind any routes (meaning changing this will change this behavior):
https://github.com/hashicorp/consul/blob/a3bf847208797462fc44c333dadc95b34163d7d2/agent/consul/gateways/controller_gateways.go#L673-L682

What this currently means is that if any of our certificate refs are invalid across any listener, we actually say that the Gateway can't bind a route, that's probably wrong and should probably change and be scoped to a given listener (which the method that this check is contained in actually has access to).

If we want to fix it, rather than mark the entire Gateway as not `Accepted`, we could potentially mark each Listener individually as `Accepted` and scope the above binding check to a listener. That said, that would potentially defeat the purpose of considering `Accepted` as a rolled-up status since I can't see how else an _entire_ Gateway would ever be `Accepted == False`.

So, the TLDR; on my initial thoughts are:

**keep this `Accepted`, switch `invalidCertificate` to `ResolvedRefs`, at least for this PR**

but I'm definitely open to discussion.

@mikemorris any thoughts?

Reply:
that makes sense to me, if anything we can make changes in a future PR once we get a better feel for how this should look